apps/typegpu-docs/src/content/examples/simulation/confetti/index.ts

```diff
@@ -0,0 +1,256 @@
+  'instance',
+);
+
+const dataUniform = asMutable(particleDataBuffer);
+const deltaTimeUniform = asUniform(deltaTimeBuffer);
+const timeStorage = asMutable(timeBuffer);
+
+// functions
+
+const rotate = tgpu.fn([vec2f, f32], vec2f).does(/* wgsl */ `
+  (v: vec2f, angle: f32) -> vec2f {
+    let pos = vec2(
+      (v.x * cos(angle)) - (v.y * sin(angle)),
+      (v.x * sin(angle)) + (v.y * cos(angle))
+    );
+
+    return pos;
+  }
+`);
+
+const mainVert = tgpu
+  .vertexFn(
+    {
+      tilt: f32,
+      angle: f32,
+      color: vec4f,
+      center: vec2f,
+      index: builtin.vertexIndex,
+    },
+    VertexOutput,
+  )
+  .does(
+    /* wgsl */ `(
+      @location(0) tilt: f32,
+      @location(1) angle: f32,
+      @location(2) color: vec4f,
+      @location(3) center: vec2f,
+      @builtin(vertex_index) index: u32
+    ) -> VertexOutput {
+    let width = tilt;
+    let height = tilt / 2;
+
+    var pos = rotate(array<vec2f, 6>(
+      vec2f(0, 0),
+      vec2f(width, 0),
+      vec2f(width, height),
+
+      vec2f(width, height),
+      vec2f(0, height),
+      vec2f(0, 0),
+    )[index] / 350, angle) + center;
+
+    if (canvasAspectRatio < 1) {
+      pos.x /= canvasAspectRatio;
+    } else {
+      pos.y *= canvasAspectRatio;
+    }
+
+    return VertexOutput(vec4f(pos, 0.0, 1.0), color);
+  }`,
+  )
+  .$uses({
+    rotate,
+    canvasAspectRatio: canvasAspectRatioData,
+    get VertexOutput() {
+      return mainVert.Output;
+    },
+  });
+
+const mainFrag = tgpu.fragmentFn(VertexOutput, vec4f).does(/* wgsl */ `
+  (@location(0) color: vec4f) -> @location(0) vec4f {
+    return color;
+  }`);
+
+const mainCompute = tgpu
+  .computeFn([1])
+  .does(
+    /* wgsl */ `(@builtin(global_invocation_id) gid: vec3u) {
+    let index = gid.x;
+    if index == 0 {
+      time += deltaTime;
+    }
+    let phase = (time + data[index].seed) / 200; 
+    data[index].position += data[index].velocity * deltaTime / 20 + vec2f(sin(phase) / 600, cos(phase) / 500);
+  }`,
+  )
+  .$uses({ data: dataUniform, deltaTime: deltaTimeUniform, time: timeStorage });
+
+// pipelines
+
+const renderPipeline = root
+  .withVertex(mainVert, {
+    tilt: geometryLayout.attrib.tilt,
+    angle: geometryLayout.attrib.angle,
+    color: geometryLayout.attrib.color,
+    center: dataLayout.attrib.position,
+  })
+  .withFragment(mainFrag, {
+    format: presentationFormat,
+  })
+  .createPipeline()
+  .with(geometryLayout, particleGeometryBuffer)
+  .with(dataLayout, particleDataBuffer);
+
+const computePipeline = root.withCompute(mainCompute).createPipeline();
+
+// compute and draw
+
+const randomizePositions = () =>
+  particleDataBuffer.write(
+    Array(PARTICLE_AMOUNT)
+      .fill(0)
+      .map(() => ({
+        position: vec2f(Math.random() * 2 - 1, Math.random() * 2 + 1),
+        velocity: vec2f(
+          (Math.random() * 2 - 1) / 50,
+          -(Math.random() / 25 + 0.01),
+        ),
+        seed: Math.random(),
+      })),
+  );
```

Comment:
Tiny nitpick, visual preference 👁️.
```suggestion
function randomizePositions() {
  particleDataBuffer.write(
    Array(PARTICLE_AMOUNT)
      .fill(0)
      .map(() => ({
        position: vec2f(Math.random() * 2 - 1, Math.random() * 2 + 1),
        velocity: vec2f(
          (Math.random() * 2 - 1) / 50,
          -(Math.random() / 25 + 0.01),
        ),
        seed: Math.random(),
      })),
  );
}
```